### ISSUE TITLE
[BUG] Upgrade: agent nodes turn into NotReady during upgrading Harvester chart

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#1985]
**Describe the bug**

During bumping Harvester ManagedChart, I see the agent node becomes NotReady.

rke2-agent log of agent node:
```
Mar 08 15:20:16 node3 rke2[2191]: Flag --resolv-conf has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/t>
Mar 08 15:20:16 node3 rke2[2191]: Flag --serialize-image-pulls has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes>
Mar 08 15:20:16 node3 rke2[2191]: Flag --tls-cert-file has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs>
Mar 08 15:20:16 node3 rke2[2191]: Flag --tls-private-key-file has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.>
Mar 08 15:20:21 node3 rke2[2029]: time="2022-03-08T15:20:21Z" level=info msg="Stopped tunnel to 10.10.0.100:9345"
Mar 08 15:20:21 node3 rke2[2029]: time="2022-03-08T15:20:21Z" level=info msg="Connecting to proxy" url="wss://10.10.0.12:9345/v1-rke2/connect"
Mar 08 15:20:21 node3 rke2[2029]: time="2022-03-08T15:20:21Z" level=info msg="Proxy done" err="context canceled" url="wss://10.10.0.100:9345/v1-rke2/connect"
Mar 08 15:20:21 node3 rke2[2029]: time="2022-03-08T15:20:21Z" level=info msg="Connecting to proxy" url="wss://10.10.0.14:9345/v1-rke2/connect"
Mar 08 15:20:21 node3 rke2[2029]: time="2022-03-08T15:20:21Z" level=info msg="Connecting to proxy" url="wss://10.10.0.11:9345/v1-rke2/connect"
```

**To Reproduce**
1. Create a 1.0.0 cluster.
2. Upgrade with master ISO.

** More context **

It might be VIP temporarily unavailable during the upgrade.

**Comment by FrankYang0529** [comment]
I can't reproduce this on a 2-node cluster (1 master node and 1 agent node) from `v1.0.0` to `master-60b2fe4-head`.

**Comment by bk201** [comment]
This is related to https://github.com/harvester/harvester/issues/1521 
The reason that the agent node goes into NotReady is because VIP moves to another node during upgrade.

I have 4 node cluster (3 server and 1 agent).
If I start them together, the agent nodes' kubelet config will looks like (10.10.0.100 is the VIP)

```
# cat /var/lib/rancher/rke2/agent/kubelet.kubeconfig
apiVersion: v1
clusters:
- cluster:
    server: https://10.10.0.100:6443

...
```

If start server nodes first, and start agent node later. Kubelet on agent node will use local loadbalancer correctly:
And in this config, moving VIP doesn't fail the agent node.
```
node3:~ # cat /var/lib/rancher/rke2/agent/kubelet.kubeconfig
apiVersion: v1
clusters:
- cluster:
    server: https://127.0.0.1:6443
    certificate-authority: /var/lib/rancher/rke2/agent/server-ca.crt
...
```

**Comment by guangbochen** [comment]
We might be able to patch the agent node server URL before the upgrade as a workaround. 
Additionally, wondering if we can leverage Rancherd's [Server discovery](https://github.com/rancher/rancherd#server-discovery) feature of the PXE boot installation with node >=  3.

**Comment by bk201** [comment]
@guangbochen The agent node server URL [is already patched with VIP in v1.0.0.](https://github.com/harvester/harvester-installer/blob/master/package/harvester-os/files/usr/sbin/harv-update-rke2-server-url)
But if an agent node doesn't wait for server nodes to be ready, it seems to use the VIP instead of local LB as its kubelet server URL.

**Comment by bk201** [comment]
Thanks to @FrankYang0529 and @yaocw2020, we've found the VIP doesn't function very well during an upgrade if the VIP moved to another node.

Here is what we see:

### Prerequisities
Agent node must start **before** server nodes and its kubelet **doesn't** use local load balancer (see https://github.com/harvester/harvester/issues/1985#issuecomment-1066736221, the `server` field must use VIP in its URL).

### Procedures
Upgrade harvester and VIP might moved
- On the agent node, we can ping VIP without issue, but we can't curl kube-api
  ```
  node3:~ # curl https://10.10.0.100:6443/healthz --cacert /var/lib/rancher/rke2/agent/server-ca.crt --cert /var/lib/rancher/rke2/agent/client-kubelet.crt --key /var/lib/rancher/rke2/agent/client-kubelet.key
  ## will block
  ```

- On all other 3 server nodes, curling kube-api works.
- If we do this on agent node, agent node can reach kube-api again:
  ```
  $ arping -I harvester-mgmt 10.10.0.100
  ```

- The MAC recognized by the agent node changes after issuing the arping command to the VIP:
  ```
  # before
  node3:/home/rancher # ip neigh | grep 10.10.0.100
  10.10.0.100 dev harvester-mgmt lladdr 52:54:00:6a:59:7c REACHABLE
  
  # after
  node3:/home/rancher # ip neigh | grep 10.10.0.100
  10.10.0.100 dev harvester-mgmt lladdr 52:54:00:84:ef:b9 REACHABLE
  ```

**Comment by yaocw2020** [comment]
The reason is that the agent node missed the gratuitous arp which is sent by the kube-vip only one time when transferring from one management host to another.

**Comment by yaocw2020** [comment]
Blocked by https://github.com/kube-vip/kube-vip/issues/382

**Comment by TachunLin** [comment]
Verified **fixed** on **`v1.0.1-rc3`**. Close this issue.

### Result
* Can 100% upgrade 4 nodes harvester (3 management, 1 agent node) from v1.0.0 to v1.0.1-rc3
  ![image](https://user-images.githubusercontent.com/29251855/162408820-e2540958-f436-48e3-a7b1-31992b788e7c.png)
* While `Upgrade system services`, none of the agent node enter in to `NotReady` state
  ![image](https://user-images.githubusercontent.com/29251855/162387442-539e61b2-93f8-4f01-b41d-3798332e5a92.png)

### Test Information
* Test Environment: 4 nodes harvester on dolphin bare machines
* Harvester version: v1.0.0 GA to v1.0.1-rc3

### Verify Steps
1. Prepare v1.0.0 four nodes harvester cluster on Provo machines
2. Create a ubuntu image
3. Create a ubuntu virtual machine
4. Enable network to `harvester-mgmt`
5. Shutdown all VMs
6. Follow the steps to upgrade to v1.0.1-rc3
https://github.com/harvester/docs/blob/42598035cfe79856f7e159d4c47c4d7d29178900/docs/upgrade/automatic.md
7. Check the node status while in the `Upgrade System Services`

### Additional Context
* Support Bundle
[supportbundle_800060cb-051a-47cd-8bff-7cb1ae5090aa_2022-04-08T10-17-57Z.zip](https://github.com/harvester/harvester/files/8450956/supportbundle_800060cb-051a-47cd-8bff-7cb1ae5090aa_2022-04-08T10-17-57Z.zip)
